UDP Timeout text (#222)
quicwg/ops-drafts: draft-ietf-quic-manageability.md

and observation of the handshake for connection confirmation ({{sec-confirm}}).
The lack of any visible end-of-flow signal ({{sec-teardown}}) means that this
state must be purged either through timers or through least-recently-used
eviction, depending on application requirements.

{{?RFC4787}} requires a timeout that is not less than 2 minutes for most UDP
traffic.  However, in practice, timers are sometimes lower, in the range of 30
to 60 seconds. In contrast, {{?RFC5382}} recommends a timeout of more than 2
hours for TCP, given that TCP is a connection-oriented protocol with well-
defined closure semantics.

As QUIC has features to improve robustness to NAT rebinding, devices might be
tempted to further reduce the UDP timeout to avoid unnecessary state. However,
this has two drawbacks.

First, not all endpoints use routing architectures where connections will
survive a port or address change. Second, applications which allow either peer
to send after a period of quiescence will experience black-holing until the peer
re-initializes the NAT binding. For these reasons, retaining the {{?RFC4787}}
limits is useful, even where devices are able to distinguish QUIC traffic from

> other UDP payloads.

Reviewer: Not sure if the advise is crystal clear here now. Initially the advise was more like "even if you use too short time-outs for UDP, please please don't do that for QUIC and use at least two minutes". While now the recommendation here rather reads like "just do the same thing as you do for UDP".

Reviewer 2: I think I agree, so can we change this: "For these reasons, retaining the {{?RFC4787}} limits is useful, even where devices are able to distinguish QUIC traffic from other UDP payloads." to something stronger, e.g. perhaps like:
"For these reasons, the limits in {{?RFC4787}} are important to avoid black-holing of packets (and hence avoid interrupting the flow of data to the client), even where devices are able to distinguish QUIC traffic from other UDP payloads."

> As QUIC has features to improve robustness to NAT rebinding, devices might be
> tempted to further reduce the UDP timeout to avoid unnecessary state. However,
> this has two drawbacks.
> 
> First, not all endpoints use routing architectures where connections will
> survive a port or address change. Second, applications which allow either peer
> to send after a period of quiescence will experience black-holing until the peer
> re-initializes the NAT binding. For these reasons, retaining the {{?RFC4787}}
> limits is useful, even where devices are able to distinguish QUIC traffic from
> other UDP payloads.

Reviewer: ```suggestion
Even though QUIC has explicitly been designed to improve robustness to NAT rebindings,
which might look like a tempting feature to further reduce the UDP timeout and thereby
minimize state, this is not recommended.  Instead it is recommended, even when lower
timers are used for other UDP traffic, to use a timer of at least two minutes for QUIC traffic.

While QUIC has no clear network-visible end-of-connection signal and therefore
does require timer-based state removal, the QUIC handshake indicates confirmation
of both ends that a valid bidirectional transmission is on-going. If state is removed
too early, this could lead to black-holing of incoming packets after a short idle period.
To detect this situation, a timer at the client needs to expire before a re-establishment
can happen (if at all), which would lead to unnecessary long delays in an otherwise
working connection.

Furrther, not all endpoints use routing architectures where connections will
survive a port or address change. So even when the client revives the connection,
a NAT rebinding can cause a routing mismatch where a packet is not even delivered to
the server that might support address migration. 

For these reasons, the limits in {{?RFC4787}} are important to avoid black-holing
of packets (and hence avoid interrupting the flow of data to the client), especially where
devices are able to distinguish QUIC traffic from other UDP payloads.
```

Just wanted to make a proposal to make the recommendation a bit stronger but then ended up with quite a bit a rewrite... so here is my proposal...

Reviewer 2: Looks better to me!

Check: /Furrther,/Further,/ or Furthermore,

timers or through least-recently-used eviction, depending on application
requirements.

{{?RFC4787}} requires a timeout that is not less than 2 minutes for most UDP

> traffic.  However, in pratice, timers are often lower, in the range of 15 to 30

Reviewer 3: 15 seconds is very rare unless something has changed recently, so I'd say 30 seconds to a minute.

Reviewer 4: OK, I just took that from the original text. Happy to change it.

Reviewer 3: Understood, but yes, let's change it.

Reviewer: I thought we took this from some paper but can't real find it anymore. I found this one though which actually indicates that most timeout are 2 minutes: https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.739.5004&rep=rep1&type=pdf

> traffic.  However, in pratice, timers are sometimes lower, in the range of 30 to
> 60 seconds. In contrast, {{?RFC5382}} recommends a timeout of more than 2 hours

Reviewer 5: ```suggestion
traffic.  However, in practice, timers are sometimes lower, in the range of 30 to
60 seconds. In contrast, {{?RFC5382}} recommends a timeout of more than 2 hours
```

TCP, given that TCP is a connection-oriented protocol with well-defined closure
semantics.

As QUIC has features to improve robustness to NAT rebinding, and has no

> observable end-of-connection semantics, devices might choose to further reduce

Reviewer 3: Most UDP protocols don't have end-of-connection semantics, so this comparison seems odd in this context, since the UDP timeout is already much less than TCP.

Reviewer 3: ```suggestion
observable end-of-connection semantics, devices might be tempted to further reduce
```

Reviewer: I'd recommend to rather start this paragraph with saying what we recommend and not saying what we no recommend :-)

> As QUIC has features to improve robustness to NAT rebinding, and has no
> observable end-of-connection semantics, devices might choose to further reduce
> the UDP timeout to avoid unnecessary state. However, this has two drawbacks.

Reviewer 3: ```suggestion
As QUIC has features to improve robustness to NAT rebinding, devices might be
tempted to further reduce the UDP timeout to avoid unnecessary state. However,
this has two drawbacks.
```

I don't like 'tempted', but the current text read to me as though you're saying it's ok, which I do not want to convey.

Reviewer 4: accepted this manually

> First, not all servers inhabit routing architectures where connections will

Reviewer 5: ```suggestion
First, not all endpoints use routing architectures where connections will
```

Not sure about endpoints vs. servers here, but "inhabit" implies life.

endpoint to re-instantiate the connection for security reasons, further

> reducing the likelihood of successful revival. For these reasons, devices that

Reviewer 3: I don't think I understand this second point about re-instantiating the connection for security reason.

Reviewer 4: This is the thing where a server packet does no good for rebinding because the firewall expects the first packet from the client. Can you suggest alternate text?

Reviewer 3: Suggestion below.

> survive a port or address change. Second, an intermediary might require one
> endpoint to re-instantiate the connection for security reasons, further
> reducing the likelihood of successful revival. For these reasons, devices that

Reviewer 3: ```suggestion
survive a port or address change. Second, applications which allow either peer to send after a
period of quiescence will experience black-holing until the peer re-initializes the NAT binding.
For these reasons, devices that
```

> re-initializes the NAT binding. For these reasons, devices that can distinguish
> QUIC traffic from other UDP payloads should retain the {{?RFC4787}} limits.

Reviewer 5: ```suggestion
re-initializes the NAT binding. For these reasons, retain the {{?RFC4787}} limits is
useful, even where devices are able to distinguish QUIC traffic from other UDP payloads.
```

> {{?RFC4787}} limits.

Reviewer: I know the originally text wasn't that much better, but I guess actually all devices should retain the RFC4787 limits for all UDP traffic because that's what we recommend in RFC4787. This text should only say that it even more important and also more reasonable to retain these limits for QUIC because QUIC is connection-oriented and therefore a device can at least because sure that there is a valid communication going on if it sees an successful handshake. 

I actually don't think, it a good advise to tell these boxes to be QUIC aware. It's only if you really don't want to adjust your timers for other UDP traffic than you should still consider doing so for QUIC specifically which however means you need additional logic to detect QUIC traffic, which however in itself is not super easy (see other section).
